src/patchset.py

```diff
@@ -0,0 +1,250 @@
+            )
+
+        patch_artifacts = patchset_node.get("artifacts")
+        if not patch_artifacts:
+            raise ValueError(
+                f"No patch artifacts available for node {patchset_node['id']}"
+            )
+
+        if not all(
+            self._has_approved_domain(patch_mbox_url)
+            for patch_mbox_url in patch_artifacts.values()
+        ):
+            raise RuntimeError("Forbidden domain")
+
+        patchset_build_config = self._find_build_config(patchset_node)
+        if build_config != patchset_build_config:
+            raise ValueError(
+                f"Patchsets node {patchset_node['id']} build config "
+                f"doesn't match to the parent node {checkout_node['id']}"
+            )
+
+        self._download_checkout_archive(checkout_node["artifacts"]["tarball"])
+
+        chekout_name = '-'.join([
+            'linux',
+            checkout_node["revision"]["tree"],
+            checkout_node["revision"]["branch"],
+            checkout_node["revision"]["describe"],
+        ])
+        chekout_path = os.path.join(self._kdir, chekout_name)
+        patchset_hash = self._apply_patches(chekout_path, patch_artifacts)
+        patchset_hash_short = patchset_hash[:PATCHSET_SHORT_HASH_LEN]
+        tarball_path = self._make_tarball(
+            target_dir=chekout_path,
+            tarball_name=f"{chekout_name}-{patchset_hash_short}"
+        )
+        tarball_url = self._push_tarball(tarball_path)
+
+        patchset_revision = checkout_node["revision"].copy()
+        patchset_revision['patchset'] = patchset_hash
+
+        self._update_node(
+            patchset_node,
+            patchset_revision,
+            tarball_url
+        )
+
+    def _mark_failed(self, patchset_node):
+        node = patchset_node.copy()
+        node.update({
+            'state': 'done',
+            'result': 'fail',
+        })
+        try:
+            self._api.update_node(node)
+        except requests.exceptions.HTTPError as err:
+            err_msg = json.loads(err.response.content).get("detail", [])
+            self.log.error(err_msg)
+
+    def _run(self, sub_id):
+        self.log.info("Listening for new trigger events")
+        self.log.info("Press Ctrl-C to stop.")
+
+        while True:
+            patchset_nodes = self._api.get_nodes({
+                "name": "patchset",
+                "state": "running",
+            })
+
+            for patchset_node in patchset_nodes:
+                if not patchset_node["parent"]:
+                    continue
+
+                checkout_node = self._api.get_node(patchset_node["parent"])
+                if checkout_node["state"] == 'running':
+                    continue
+
+                try:
+                    self._process_patchset(checkout_node, patchset_node)
+                except Exception as e:
+                    self._mark_failed(patchset_node)
+                    self.log.error(e)
+
+            time.sleep(10)
```

Comment:
This service currently implements equivalent of `trigger` and extension to `tarball` - is it intended to be divided further with a separate trigger service? I don't suggest doing it right now, I'm rather curious if [trigger-like options](https://github.com/kernelci/kernelci-pipeline/blob/9e68e16e11c683af463aeb1d18f0ea46ff6b1b5a/src/trigger.py#L84) would also apply here

Reply:
The biggest difficulty I faced with this logic was the condition that we need to start `patchset` node processing after it's parent `checkout` node becomes `available` (otherwise no source tarball available for download).

If there's a better way to do this delayed processing I am all for it.

Reply:
But, why `sleep(10)` anyway?

Reply:
Ah I see, to periodically check if a new patchset is available.  I thought there would be a bridge service or some kind of callback to get a Pub/Sub event generated from Patchwork when that happened.

Reply:
I wish there would be a way to get event, and then reschedule it if `checkout_node` isn't available yet, but it doesn't seem to exist now, so I have to poll nodes from API.

Reply:
we are in development phase, so maybe we can introduce something, to make it event driven. For example we can introduce pipeline service that will trigger external callback and provide to external URL checkout_node details.

Reply:
(will proceed with current implementation, adding a TODO to allow event-driven triggering for patchsets)